rp: add new pio spi program

wowkster opened the pull request:
# Summary

This PR adds a new PIO based SPI driver to ``embassy-rp``. It supports both blocking and async operation using DMA transfers just like the hardware SPI driver, but currently can not be reconfigured at runtime without making breaking changes to the PIO APIs. 

# API Differences

I attempted to make the interface as similar as possible to the current hardware driver, but some changes were necessary to make configuration simpler. The biggest difference is that the PIO program has to be configured with the clock phase, since it isn't possible to use the same PIO program for CPHA0 and CPHA1 modes. Because of this limitation, I decided to create a separate ``Config`` structure for the PIO SPI driver which omits the phase.

# Unfinished Business

I think it would be possible to reconfigure the clock polarity and frequency at runtime by changing the state machine configuration, but changing the clock phase would be challenging. Currently there is a mechanism to free PIO instruction memory, but ownership of that memory is not tracked through the type system since ``LoadedProgram`` references are not kept after configuration.

# Open Questions

To make the async implementation work, I had to do what is akin to ``dma::read_repeated`` if it existed. Is it worth creating this function for situations where data needs to be discarded through DMA?

Additionally, The PIO implementation currently sets up its own DMA transfers since it needs access to the DMA ``CTRL_TRIG.BSWAP`` bit. Could these be refactored to use the DMA helper functions? The BSWAP bit could be set to the required value and then restored after the transfer is complete.

wowkster committed:
rp: add new pio dma apis

This commit adds StateMachineRx::dma_pull_repeated and
StateMachineTx::dma_push_repeated which allow you to discard reads or
send dummy writes to the state machine using the DMA hardware

wowkster committed:
rp: add new pio spi program

This commit adds a new PIO program which implements SPI. This allows
you to drive more than 2 SPI buses by using PIO state machines as
additional duplex SPI interfaces.

The driver supports both blocking and async modes of operation and
exclusively uses the DMA for async IO.

wowkster committed:
rp: add pio spi examples

wowkster committed:
style: cleanup with rustfmt

CBJamo commented:
This is great! I'm super happy to see more pio programs getting added. I am curious about what you need more than 2 spi ports for, if you can talk about it publicly.

Feel free to change/refine LoadedProgram, it was just the most obvious way allow configuring multiple SMs without duplicating the program. FYI, I think you're the first person to implement any new pio programs in embassy-rp itself after I moved/normalized the examples.

I feel the same about the broader pio api. If changes are needed to expose more complete functionality then make them. I don't think anyone did super intensive design here, so we got here by adding features as new programs need them. Please correct me if that design did happen somewhere (probably matrix?) and I missed it.

wowkster commented:
Thank you! I found it interesting to implement and would love to answer your questions. As to why we needed more than 2 SPI ports, we are developing safety critical motion control hardware using the RP235xy and require high SPI bandwidth to simultaneously interact with a CAN FD controller, read motor status and encoder data from multiple TMC2240 stepper motor drivers, and read sensor feedback data from an SPI compatible high precision ADC, all with low latency. Given the different mode and frequency requirements of these devices and our ADC's specific quirks which make it difficult to use through the SpiDevice API, we found it simpler to allocate extra GPIO pins to a dedicated PIO driven SPI bus.

I think the LoadedProgram API is good since it allows sharing an installation across multiple state machines but it breaks down slightly when needing to reconfigure the state machines. Through an exclusive reference to the LoadedProgram, it would be safe to unload it and attempt to load a new altered version. However, without references being stored by the drivers after initialization, designing an API that isn't cumbersome to use would present some challenges. 

I propose for some of the PIO drivers to take ownership of their programs if the user wants to reconfigure them at runtime. It would cause some duplication, but in practice I think it would be minimal. Additionally, just like many peripherals use a type state parameter to track whether or not they own a DMA channel, maybe certain PIO drivers could use a type state parameter to keep track of whether they own the PIO instruction memory and can be reconfigured, or if they only hold a shared reference to it. This allows the same memory to be shared while also solving the issue of freeing instruction memory while the state machine is still running. If access to the LoadedProgram is not actually required when holding a reference, drivers could hold a phantom reference instead to make the abstraction zero-cost.

With the above changes, the PIO SPI driver could optionally support initial configuration as well as runtime reconfiguration in the same way that the SPI peripheral driver does.

My questions about the DMA APIs still hold but I recognize that this may not be the best place to address them.

Dirbaio commented:
bender run

CBJamo commented:
> ADC's specific quirks 

Why is it always an ADC where you end up needing dedicated or non-standard spi? Sounds like a cool project though!

> I propose for some of the PIO drivers to take ownership of their programs if the user wants to reconfigure them at runtime. 

In this case, since the program is either 2 or 3 instructions, I think we should skip the LoadedProgram step entirely. Just give the `Spi` struct exclusive ownership of 3 instruction slots and install the appropriate program for the polarity you need at runtime. In the small number of cases where a user needs the "wasted" slot, all the pio stuff is still exposed and they can do it by hand.

I also like your PhantomData idea, but it just doesn't feel like it's worth the complexity here with such small pio programs.

wowkster committed:
rp: add pio spi runtime reconfiguration

wowkster committed:
rp: fix pio spi async example

wowkster commented:
Do these changes look sufficient? I added runtime reconfiguration APIs similar to the builtin SPI peripheral. The only thing that I couldn't implement was the `SetConfig` trait from `embassy_embedded_hal` since reloading the program requires exclusive access to the PIO handle.

wowkster commented:
At slower SPI speeds, that may block for a sufficiently long period which could be undesirable.  I think the configuration change is sensible and I don't know that one approach is necessary better than the other. Personally I would err on the side of using less memory if it's possible.

CBJamo commented:
If we loaded both programs and put them in a `SpiPioProgram` struct, then that struct could be shared between multiple SMs. Compared to the previous plan we're only down 2 instructions, and it would come out ahead as soon as there was a second SM configured for SPI. Since it would also allow for `SetConfig` that seems like a good tradeoff to me. 

I agree that DMA is essential for SPI. In some applications (sd cards being the most common) there are large and fast transfers. In those cases the SM's fifo is too small, and it's going fast enough you don't want the cpu to get involved.

CBJamo commented:
I think it's the best design we've discussed so far. If we ever come up with a better one we can always have a follow on PR.

wowkster commented:
Ok I will get that in ASAP

wowkster committed:
rp: move pio pin configs after set_config

This is needed for the program to work correctly on rp235xb when using
the higher pin numbers.

wowkster committed:
rp: read pio gpiobase in set_pins and set_pin_dirs
